File: php/perfect-numbers/PerfectNumbers.php
<?php

declare(strict_types=1);

function getClassification(int $number): string
{
    $sum = aliquotSum($number);
    if ($sum < $number) return 'deficient';
    if ($sum > $number) return 'abundant';
    return 'perfect';
}

function aliquotSum(int $n): int
{
    if ($n < 1) throw new InvalidArgumentException();

    $factors = [];
    for ($i = 1; $i <= sqrt($n); $i++) {
        if ($n % $i == 0) {
            $factors[$i] = true;
            $factors[$n / $i] = true;
        }
    }
    unset($factors[$n]);
    return array_sum(array_keys($factors));
}

File: euphoria/perfect-numbers/perfect-numbers.ex
type positive_integer(atom number)
  return integer(number) and number > 0
end type

public function classify(atom n)
  if not positive_integer(n) then
    return 0
  end if

  switch compare(aliquot_sum(n), n) do
    case -1 then return "deficient"
    case  0 then return "perfect"
    case  1 then return "abundant"
  end switch
end function

function aliquot_sum(atom n)
  atom sum = 0
  for f = 1 to floor(sqrt(n)) do
    if remainder(n, f) = 0 then
      sum += f
      if f != n / f then
        sum += n / f
      end if
    end if
  end for
  return sum - n
end function
